### PR TITLE
Adding arg parser files

### PR DESCRIPTION
**Pull request by nader-zouaoui**
## Description
In this pull request 2 new classes were introduced:

- arg_parser_arg: this class is an abstract representation of what a flag may be and what it needs to be fully defined and used throuought the project. It exports 2 types of arguments
  - arg_parser_arg_opt: an optional command line flag that takes no additional arguments
  - arg_parser_arg_pos: a positional argument that expects at least one additional argument 
    - PS: if you pass -1 as it's arg_count that it takes on as many arguments as you give it.
- arg_parser: This class is where we define all the command line arguments that our tool supports.
  - it implements multiple instances of the arg_parser_arg class and contains the logic for parsing all of the defined args
  - it also contains the print help method that prints to the console all the information related to the commands defined in our tool
## Related Issue
related to #35 
## Commits

${{ github.event.pull_request.commits }}

## How Has This Been Tested?

image (7)

### CODE REVIEW

File: wperf/arg-parser-arg.cpp [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+            {
+                formatted_str += current_line + L"\n\n";
+            }
+        }
+
+        // Remove the trailing newline, if any
+        if (!formatted_str.empty() && formatted_str.back() == L'\n')
+            formatted_str.pop_back();
+        if (!formatted_str.empty() && formatted_str.back() == L'\n')
+            formatted_str.pop_back();
```
Comment (PrzemekWirkus) [inline wperf/arg-parser-arg.cpp:215]: Is this code doubled on purpose?

File: wperf-test/arg-parser-arg-tests.cpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+    TEST_CLASS(ArgParserArgTests)
+    {
+    public:
+        TEST_METHOD(TestConstructor)
+        {
+            arg_parser_arg arg(L"--name", { L"-n" }, L"Description of the argument", { L"default" }, 1);
+            Assert::AreEqual(std::wstring(L"--name"), arg.get_name());
+            Assert::AreEqual(std::wstring(L"-n"), arg.get_alias_string());
+            Assert::AreEqual(std::wstring(L"Description of the argument"), arg.get_usage_text());
```
Comment (PrzemekWirkus) [inline wperf-test/arg-parser-arg-tests.cpp:50]: Dunno what `L"default"` is but you are not testing for it.

File: wperf-test/arg-parser-arg-tests.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+            Assert::AreEqual(std::wstring(L"input.txt"), values[0]);
+            Assert::IsTrue(arg.is_set());
+        }
+        TEST_METHOD(TestGetHelpSingleAlias)
+        {
+            arg_parser_arg arg(L"--help", { L"-h" }, L"Displays help information.");
+            auto help_text = arg.get_help();
+
+            // Check if the help text includes all flags and description
+            Assert::IsTrue(help_text.find(L"--help, -h") != std::wstring::npos);
```
Comment (PrzemekWirkus) [inline wperf-test/arg-parser-arg-tests.cpp:126]: To e.g. test if that extra comma is not at the end...
```
Assert::IsFalse(help_text.find(L"--help, -h,") != std::wstring::npos);
```

File: wperf/arg-parser-arg.cpp [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    {
+        return m_arg_count;
+    }
+
+    std::wstring arg_parser_arg::get_name() const
+    {
+        return m_name;
+    }
+
+    std::wstring arg_parser_arg::get_alias_string() const
```
Comment (PrzemekWirkus) [inline wperf/arg-parser-arg.cpp:86]: Maybe you can reuse:
```cpp
std::wstring WStringJoin(const std::vector<std::wstring>& input, std::wstring sep)
```

from `utils.cpp` ?

File: wperf/arg-parser-arg.cpp [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    {
+        return m_values;
+    }
+
+    bool arg_parser_arg::parse(std::vector<std::wstring> arg_vect)
+    {
+        if (arg_vect.size() == 0 || !is_match(arg_vect[0]))
+            return false;
+
+        if (m_arg_count == -1 && arg_vect.size() > 0) m_arg_count = arg_vect.size() - 1;
```
Comment (PrzemekWirkus) [inline wperf/arg-parser-arg.cpp:130]: Please do not use `-1` with `size_t`, which is unsigned type.

File: wperf/arg-parser-arg.cpp [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        if (arg_vect.size() < m_arg_count + 1)
+            throw std::invalid_argument("Not enough arguments provided.");
+
+        if (m_arg_count == 0)
+        {
+            set_is_parsed();
+            return true;
+        }
+
+        for (int i = 1; i < m_arg_count + 1; ++i)
```
Comment (PrzemekWirkus) [inline wperf/arg-parser-arg.cpp:141]: Use one type for index `i`.

File: wperf/arg-parser-arg.cpp [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+                    throw std::invalid_argument("Invalid arguments provided.");
+            }
+            m_values.push_back(arg_vect[i]);
+        }
+        set_is_parsed();
+        return true;
+    }
+
+
+    std::wstring arg_parser_add_wstring_behind_multiline_text(const std::wstring& str, const std::wstring& prefix)
```
Comment (PrzemekWirkus) [inline wperf/arg-parser-arg.cpp:155]:
```cpp
// At least 1 line of text what this function should do.
```

File: wperf/arg-parser-arg.cpp [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+                formatted_str += L"\n";
+                continue;
+            }
+
+            formatted_str += prefix + current_line + L"\n";
+        }
+        return formatted_str;
+    }
+
+    std::wstring arg_parser_format_string_to_length(const std::wstring& str, size_t max_width)
```
Comment (PrzemekWirkus) [inline wperf/arg-parser-arg.cpp:173]:
```cpp
// At least 1 line of text what this function should do.
```

Comment (nader-zouaoui) [reply wperf/arg-parser-arg.cpp:173]: That is represented in the `arg-parser-arg-utils-tests` file

File: wperf/arg-parser-arg.cpp (hunk 4 above)
Comment (nader-zouaoui) [reply wperf/arg-parser-arg.cpp:87]: Done

File: wperf-test/arg-parser-arg-utils-tests.cpp [hunk 9]
```diff
@@ -62,26 +62,37 @@ namespace arg_parser_arg_utils_tests
         {
             std::wstring input = L"This is a line that exceeds the width.";
             size_t max_width = 15;
-            std::wstring expected = L"This is a line\nthat exceeds\nthe width.";
+            std::wstring expected = L"This is a line\nthat exceeds the\nwidth.";
```
Comment (PrzemekWirkus) [inline wperf-test/arg-parser-arg-utils-tests.cpp:65]: This test is showing your funtion is not working correctly?
```
that exceeds the
1234512345123456
```
There are 16 characters in second line, is that correct if `max_width` is `15`?

**Review by PrzemekWirkus — CHANGES_REQUESTED** [review]

**Review by PrzemekWirkus — APPROVED** [review]

**Pull request merged**